### PR TITLE
Form types for Django 4.0.

### PR DESCRIPTION
**Pull request by smithdc1**
Work in progress to add type for changes to Django forms introduced in 4.0

### CODE REVIEW

File: django-stubs/forms/utils.pyi [hunk 1]
```diff
@@ -13,26 +15,54 @@ _FilesT = MultiValueDict[str, UploadedFile]
...
+        renderer: Optional[BaseRenderer] = ...,
+    ) -> None: ...
     def as_data(self) -> Dict[str, List[ValidationError]]: ...
     def get_json_data(self, escape_html: bool = ...) -> Dict[str, Any]: ...
-    def as_json(self, escape_html: bool = ...) -> str: ...
-    def as_ul(self) -> str: ...
-    def as_text(self) -> str: ...
 
-class ErrorList(UserList):
+class ErrorList(UserList, List, RenderableErrorMixin):
```
Comment (smithdc1) [inline django-stubs/forms/utils.pyi]: I expect the tests to fail here, as this class now inherits from both `UserList` and `List`. 

https://github.com/django/django/blob/90d2f9f41671ef01c8e8e7b5648f95c9bf512aae/django/forms/utils.py#L138

🤷

File: django-stubs/forms/renderers.pyi [hunk 2]
```diff
@@ -9,6 +9,8 @@ from django.template.base import Template
 def get_default_renderer() -> BaseRenderer: ...
 
 class BaseRenderer:
+    form_template_name: str = ...
+    formset_template_name: str = ...
```
Comment (smithdc1) [inline django-stubs/forms/renderers.pyi]: So these are some of the 4.1 changes. In 4.0 they are on `BaseForm`.

I'm not sure which version is targeted at any given time given the project does not aim to support multiple versions of Django at once?

**Review by smithdc1 — COMMENTED** [review]
Couple of questions / comments where any input would be much appreciated!

File: django-stubs/forms/renderers.pyi (hunk 2 above)
Comment (sobolevn) [reply django-stubs/forms/renderers.pyi]: We do not target 4.1 yet.

File: django-stubs/forms/utils.pyi (hunk 1 above)
Comment (sobolevn) [reply django-stubs/forms/utils.pyi]: `List[Something]` might be good enough.

Comment (smithdc1) [reply django-stubs/forms/utils.pyi]: Looks like that will do. Thank you!

File: django-stubs/forms/renderers.pyi (hunk 2 above)
Comment (smithdc1) [reply django-stubs/forms/renderers.pyi]: Thanks -- I've rolled back the 4.1 changes. This now targets 4.0.

File: django-stubs/forms/utils.pyi [hunk 3]
```diff
@@ -49,7 +48,7 @@ class ErrorDict(dict, RenderableErrorMixin):
     def as_data(self) -> Dict[str, List[ValidationError]]: ...
     def get_json_data(self, escape_html: bool = ...) -> Dict[str, Any]: ...
 
-class ErrorList(UserList, List, RenderableErrorMixin):
+class ErrorList(List[Union[ErrorList, Sequence[Union[str, Exception]]]], RenderableErrorMixin):
```
Comment (sobolevn) [inline django-stubs/forms/utils.pyi]: It should be either `Sequence` or `List`.
List is just a mutable sequence (kind of).

**Pull request closed without merging**